### ISSUE TITLE
Perform "build artifact caching"

### ISSUE THREAD

**Issue by f-f** [purescript/spago#527]
Problem: if I'm using the library `y` from package set `x`, it's silly that I have to rebuild this in every project. We currently have some of this caching in the form of sharing the output folder in the same monorepo, but this should not happen for reasons defined in #526 

Instead of doing that, I propose we cache build products a-là Nix derivations, with some form of [content addressable cache](https://en.wikipedia.org/wiki/Content-addressable_storage).
How this would work in practice is that whenever we need to build a package, we:
- hash the `PackageSet` together with the compiler version
- lookup into the build cache if we have a folder with that hash
- if yes, we look in there and see if we built that library for these set and purs version. If yes, we copy it over to our `output` folder
- we don't find it or we don't have the folder with the hash, then we build it and put it in the cache

**Comment by hdgarrood** [comment]
You'll have to manually merge `cache-db.json` files if you want this to work, but I'd highly discourage that for a couple of reasons: firstly they are not part of the compiler's public API, so can change without warning, and secondly incremental builds are difficult enough to get right as it is, and I think having multiple different tools operating on these files makes it much more likely for things to malfunction. I think nothing other than the compiler should be writing anything inside an output directory (perhaps we could document this better on the compiler side).

I think it makes more sense to implement this inside the compiler (see: https://github.com/purescript/purescript/issues/3493, https://github.com/purescript/purescript/issues/2477).

**Comment by f-f** [comment]
@hdgarrood both the issues you linked are about making the compiler package-aware, which is something I would not like to happen at all.
If this global caching is going to be implemented compiler-side, IMO the right granularity for this caching to happen would be module-level - as it is right now - or ideally definition-level.

So, it looks like the situation is not entirely clear there, and I'd be fine with Spago doing things with the `cache-db.json` (why have JSON files if they are not going to be consumed by other tools? I thought that was the idea in the first place) if that would mean making progress on this independently from what the compiler does/plans - i.e. something I'm really not a fan of is blocking work on other work.
In short: if the compiler wants to implement this then it's good, but since that might not happen I'd like to explore the possibility of implementing this here

**Comment by hdgarrood** [comment]
Within the context of faster builds and not doing work that doesn't need doing, I think it makes a lot of sense to make the compiler package-aware; I'll try to expand a little and do my best not to derail. Essentially, my view is that it makes sense to consider two types of inputs to builds: published packages at specific versions, which don't change and which are often depended on across different projects, and project-local code, which changes all the time and isn't depended on across different projects. These should be treated differently by the build system: because project-local files are liable to change often, we should be checking their hashes and/or timestamps on every build, but there's little sense in checking whether published packages have changed since the last build. In fact, right now, the compiler has to not only check timestamps of all source from your dependencies on every build, it also has to parse all of their imports and construct a module graph including all modules from all dependencies. If we had a concept of a package database which the compiler could use to store precompiled packages, we could skip this step, and the compiler would only have to read and parse files which were local to the project being built. Personally, I don't really see a way to a build system which I consider sufficiently fast without addressing this. Also, this functionality would have to be in the compiler, because the logic which checks timestamps and hashes and constructs module graphs is in the compiler; if we wanted to implement this outside the compiler we'd have to pull that logic outside of it too, which I'm less keen on.

Re JSON files: the intention was never that something being a JSON file means that it's part of the compiler's public API - externs files are definitely not part of the compiler's public API, for instance. I think we mostly use JSON just because it's common, it makes debugging a bit easier, most people know how to use Aeson, and for many data types we already have the Aeson instances.

I get that contributing to the compiler is difficult, though, and work on the compiler doesn't progress very quickly, so I can't blame you at all for wanting to explore this outside the compiler.

**Comment by f-f** [comment]
It looks like the compiler is indeed the better place for this to happen, so I'll close this one
